### PR TITLE
feat: Ntfy support for incident notifications added

### LINKED ISSUE #702: Add ntfy.sh support
Ntfy.sh is a library that helps send push notifications to your phone or desktop via scripts from any computer, and/or using a REST API. 

When integrated with BlueWave Uptime, it can send incidents to a smartphone running the ntfy application. 

One advantage is that it's faster than email notifications.

### PR DESCRIPTION
**Pull request by aryanp-86**
Closes #702 
### Add ntfy.sh as a New Notification Provider

This PR introduces [ntfy.sh](https://ntfy.sh/) as a new notification provider alongside existing email option.

Implementation is according to draft discussed in #702 

### Changes

- Added ntfy as a new notification type along with the config object in the notification model

```
       ...
	type: {
		type: String,
		enum: ["email", "sms", "ntfy"],
	},
	ntfyConfig: {
		type: Object,
	}
	...
```
- Implemented NtfyService class for handling ntfy.sh notifications
The NtfyService class implements ntfy.sh notification handling with support for multiple authentication methods (no-auth, user-pass, access token). It prepares notifications with customizable headers including title, priority, and tags, while handling URL sanitization and proper error logging. The service sends monitor status updates via HTTP POST requests, with the message body indicating whether a monitored service is up or down.

- Extended notification service to handle ntfy notifications
In the handleNotifications method, ntfyService being injected through the constructor, when processing notifications of type "ntfy", it delegates the notification sending to the ntfyService by calling sendNtfyNotification with the appropriate configuration and network response data.

- Updated monitor creation and configuration UI to support ntfy settings
On the UI side, a checkbox of "Notify via ntfy.sh" is shown in the incidents section which on enabling will request users to enter required fields like friendly name, topic, server url, priority and authentication mode.

Screenshot 2024-11-10 at 18-09-55 BlueWave Uptime

- Output in ntfy mobile app
WhatsApp Image 2024-11-10 at 6 12 16 PM

### Test Checklist
- [X] Basic notification works (monitor down/up triggers ntfy notification)
- [X] All auth modes work (no-auth, user/pass, access token)
- [X] Form validation and error handling works as expected
- [X] Notification settings persist after save/reload
- [X] Coexists with email notifications without conflicts

Let me know if any further refinements are required!

### CODE REVIEW

File: Server/service/ntfyService.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+			} else {
+				this.logger.error({
+					service: this.SERVICE_NAME,
+					message: "An error occurred while sending notification",
+				});
+			}
+		} catch (error) {
+			this.logger.error(`Failed to send Ntfy notification: ${error.message}`, {
+				service: this.SERVICE_NAME,
+				method: "sendNtfyNotification",
```
Comment (ajhollid) [inline Server/service/ntfyService.js:71]: Please observe the conventions for logging errors:

```
this.logger.error({
  message: error.message,
  service: <service_name>,
  method: "<method_name>",
  stack: error.stack,
  details: <more_detailed_message_here>
});
```

File: Server/service/ntfyService.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+			if (response.status === 200) {
+				this.logger.info({
+					service: this.SERVICE_NAME,
+					message: "Ntfy notification sent successfully",
+				});
+			} else {
+				this.logger.error({
+					service: this.SERVICE_NAME,
+					message: "An error occurred while sending notification",
```
Comment (ajhollid) [inline Server/service/ntfyService.js:65]: Make sure to add the method name to these logs for ease of debugging

File: Server/db/models/Notification.js [hunk 3]
```diff
@@ -8,7 +8,10 @@ const NotificationSchema = mongoose.Schema(
 		},
 		type: {
 			type: String,
-			enum: ["email", "sms"],
+			enum: ["email", "sms", "ntfy"],
+		},
+		ntfyConfig: {
+			type: Object,
```
[the remark this answers is not in the document]
Comment (ajhollid) [reply Server/db/models/Notification.js:14]: Please do define the schema here for data integrity

**Review by ajhollid — CHANGES_REQUESTED** [review]
My main concern with this PR is that there are no tests included on the back end.  Please add tests and ensure full coverage for all back end code.

Test are found in the `./Server/tests` directory.

Aside from that there are some logging conventions that need to be followed and the DB model should be more strict to ensure data integrity.  Please see comments in code review for details :+1: 

I'll leave the front end side of the review to the front end team, please request reviews from @marcelluscaio , @jennifer-gan , and @shyna92 .

Thanks for your contribution!

**Comment by gorkem-bwl** [comment]
Wanted to check here @aryanp-86 - how is the current progress for the comments above?

**Comment by ajhollid** [comment]
This branch has fallen fairly far behind dev without any activity for a couple weeks.  I'm going to close this, if you're interested in implementing this please open a new PR with tests and requested changes.

**Pull request closed without merging**